Fix: Resolve ES Module Circular Dependency Error

mvasin opened the pull request:
## The problem

The application was failing on `npm run start` with the following error:

```
% npm run start                                               

> harproxyserver@0.0.22 start
> node --loader=ts-node/esm --inspect src/harProxyServer.ts

Debugger listening on ws://127.0.0.1:9229/e417c7f7-4177-4d25-9def-c974e8675fdf
For help, see: https://nodejs.org/en/docs/inspector
(node:58777) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)
node:internal/modules/esm/loader:309
        throw new ERR_REQUIRE_CYCLE_MODULE(message);
              ^

Error [ERR_REQUIRE_CYCLE_MODULE]: Cannot require() ES Module /Users/user/code/oss/harproxyserver/src/harProxyServer.ts in a cycle.
 ```
## The root cause

- `"module": "commonjs"` in `tsconfig.json` tells Typescript to compile to CommonJS format. This means import/export would be transpiled to `require()`/`module.exports`
- `start` script in `package.json` tells NodeJS to treat files as ES Modules (a deprecated flag)
- with `import pkg from '../package.json'` in `harProxyServer.ts`,  ESM `import` syntax is used in code compiled to CommonJS, and then forcibly interpreted as ESM by the loader.

```
┌─────────────────────────────────────────────────┐
│ 1. TypeScript compiles ESM → CommonJS           │
│    import pkg from '../package.json'            │
│    becomes                                      │
│    const pkg = require('../package.json')       │
└─────────────────┬───────────────────────────────┘
                  │
                  ▼
┌─────────────────────────────────────────────────┐
│ 2. Node.js --loader=ts-node/esm                 │
│    Forces everything back to ESM interpretation │
│    require() calls are wrapped in ESM context   │
└─────────────────┬───────────────────────────────┘
                  │
                  ▼
┌─────────────────────────────────────────────────┐
│ 3. Circular Dependency Created                  │
│    - harProxyServer.ts imports package.json     │
│    - Compiled require() wrapped in ESM          │
│    - ESM loader tries to resolve as ES Module   │
│    - package.json has no "type": "module"       │
│    - Creates require(esm) cycle                 │
│                                                 │
│    ERR_REQUIRE_CYCLE_MODULE                     │
└─────────────────────────────────────────────────┘
```

- When tsconfig uses "module": "commonjs", this becomes require()
- When --loader=ts-node/esm is used, it wraps require() in ESM context
- JSON files aren't proper ES modules → circular dependency error

## The solution

ES modules are now used consistently.

1. **Updated `tsconfig.json`**
   - Changed `"module"` from `"commonjs"` to `"ES2020"` for native ES module support
   - Added `"moduleResolution": "node"` for proper dependency resolution

2. **Updated `package.json`**
   - Added `"type": "module"` to enable ES module support project-wide
   - Replaced deprecated `node --loader=ts-node/esm` with modern `tsx` in start script
   - Removed experimental loader warnings

3. **Refactored `src/harProxyServer.ts`**
   - Added dynamic `package.json` loading using `import.meta.url` and `fileURLToPath`
   - Replaced static import with runtime JSON parsing to avoid circular dependencies
   - Uses `path.join()` for cross-platform compatibility (works on Windows, macOS, Linux)

## Summary


Aspect | Before (Broken) | After (Fixed)
-- | -- | --
tsconfig module | commonjs | ES2020
package.json type | (missing) | "module"
Runtime loader | --loader=ts-node/esm (experimental) | tsx (stable)
package.json import | Static import (causes cycle) | Dynamic runtime load
Module consistency | ❌ Three different systems | ✓ Consistent ESM

The core problem was that the code was being double-transpiled: TypeScript converted ESM to CommonJS, then the loader tried to convert it back to ESM, creating a circular dependency when importing JSON files.

Warning: the export is ESM now, @yaacov let me know if this a problem.

mvasin committed:
fix: resolve circular dependency error in ES module loading

- Switch from CommonJS to ES2020 modules in tsconfig
- Add "type": "module" to package.json
- Replace ts-node/esm loader with tsx
- Refactor package.json import using import.meta.url

Fixes ERR_REQUIRE_CYCLE_MODULE error on startup

yaacov commented:
> Warning: the export is ESM now, @yaacov let me know if this a problem.

I don't know of any package/app using this lib requiring it to be cjs, so it's ok by me. 

tests fails, can you look into that ?

mvasin committed:
fix: configure Jest for ES module compatibility

- Rename jest.config.js to jest.config.cjs to preserve CommonJS syntax
- Update Jest config to use ts-jest/presets/default-esm
- Add extensionsToTreatAsEsm and useESM configuration for ts-jest
- Add moduleNameMapper to handle .js extension resolution

Fixes Jest test failures after migrating to "type": "module" in package.json.
All tests now pass successfully with ESM support.

mvasin committed:
fix: rename eslintrc to .cjs for ESM compatibility

Rename .eslintrc.js to .eslintrc.cjs to ensure ESLint config is treated
as CommonJS when "type": "module" is set in package.json.

Fixes ESLint error: "module is not defined" when running eslint command.

mvasin commented:
> > Warning: the export is ESM now, @yaacov let me know if this a problem.
> 
> I don't know of any package/app using this lib requiring it to be cjs, so it's ok by me.
> 
> tests fails, can you look into that ?

Should be fixed, please run the pipeline once again.

yaacov commented:
thanks :green_heart:
